src/firestore/app/run.ts

```diff
@@ -173,30 +203,135 @@ export class RoarRun {
    * @async
    * @param {*} trialData - An object containing trial data.
    */
-  async writeTrial(trialData: Record<string, unknown>) {
+  async writeTrial(
+    trialData: Record<string, unknown>,
+    computedScoreCallback?: (rawScores: IRawScores) => IComputedOrNormedScores,
+    normedScoreCallback?: (computedScores: IComputedOrNormedScores) => IComputedOrNormedScores,
+  ) {
     if (!this.started) {
       throw new Error('Run has not been started yet. Use the startRun method first.');
     }
+
+    // Check that the trial has all of the required reserved keys
+    if (!requiredTrialFields.every((key) => key in trialData)) {
+      throw new Error(
+        'All ROAR trials saved to Firestore must have the following reserved keys: ' +
+          `${requiredTrialFields}.` +
+          'The current trial is missing the following required keys: ' +
+          `${requiredTrialFields.filter((key) => !(key in trialData))}.`,
+      );
+    }
+
     const trialRef = doc(collection(this.runRef, 'trials'));
 
     return setDoc(trialRef, {
       ...convertTrialToFirestore(trialData),
       serverTimestamp: serverTimestamp(),
     })
       .then(() => {
-        const runScores = {
-          numAttempted: increment(1),
-          theta: trialData.theta || null,
-          thetaSE: trialData.thetaSE || null,
-        } as IRunScores;
-
-        if (trialData.correct) {
-          runScores.numCorrect = increment(1);
+        // Here we update the scores for this run. We create scores for each subtask in the task.
+        // E.g., ROAR-PA has three subtasks: FSM, LSM, and DEL. Each subtask has its own score.
+        // Conversely, ROAR-SWR has no subtasks. It's scores are stored in the 'total' score field.
+        // If no subtask is specified, the scores for the 'total' subtask will be updated.
+        const subtask = (trialData.subtask || 'total') as string;
+        const stage = trialData.assessment_stage === 'test_response' ? 'test' : 'practice';
+
+        let scoreUpdate: IScoreUpdate = {};
+        if (subtask in this.scores.raw) {
+          // Then this subtask has already been added to this run.
+          // Simply update the block's scores.
+          this.scores.raw[subtask][stage] = {
+            theta: (trialData.theta as number) || null,
+            thetaSE: (trialData.thetaSE as number) || null,
+            numAttempted: (this.scores.raw[subtask][stage]?.numAttempted || 0) + 1,
+            numCorrect: (this.scores.raw[subtask][stage]?.numCorrect || 0) + +Boolean(trialData.correct),
+            numIncorrect: (this.scores.raw[subtask][stage]?.numIncorrect || 0) + +!trialData.correct,
+          };
+
+          // And populate the score update for Firestore.
+          scoreUpdate = {
+            [`scores.raw.${subtask}.${stage}.theta`]: (trialData.theta as number) || null,
```

Comment:
just to note: current theta in legacy database is written as "thetaEstimate" instead of "theta", so we will have to change either the app code + legacy database or here to make it consistent

Reply:
Thanks. I changed it to `thetaEstimate` in the latest commit.

Reply:
thanks!